Add README.md to umbrella (#7600)
paritytech/polkadot-sdk: umbrella/README.md

## 💻 Usage

The re-exported crates are grouped under the following feature sets.

- `node`
- `runtime`
- `std`
- `experimental`
- `runtime-benchmarks`
- `runtime-full`
- `serde`
- `tuples-96`
- `try-runtime`
- `with-tracing`

Choose which ones to enable based on the needs. For example, when building a runtime, the `Cargo.toml` may contain the following. To build a runtime that can run on a wide variety of environments, you may want to opt out of `std` with `default-features = false`.

```toml
[dependencies]
polkadot-sdk = { version = "0.12.0", features = ["runtime"], default-features = false }
```

In addition to the features above, each re-exported crate is individually feature-gated. For situations where only a small subset of crates is needed, you may consider specifying the exact set of crates you need to reduce build time.

In the contrived example below, only the `frame-support` and `frame-system` crates are needed from `polkadot-sdk`.

```toml
[dependencies]
polkadot-sdk = { version = "0.12.0", features = ["frame-support", "frame-system"] }
```

> ```rust

Reviewer: It is great that this is a full e2e working example! 

What is missing is the fact that there is no connection made between `polkadot-sdk` and `polkadot-sdk-frame`, which is also exported as a part of it.

Author: I didn't know there is a `polkadot-sdk-frame`. 👼 Do you mean it's recommended to use `frame-support` and `frame-system` from the exported `polkadot-sdk-frame`?

Reviewer: You can learn about both here, and I think this should help you better structure this README to encompass both :) 

https://blog.kianenigma.com/posts/tech/polkadot-sdk-2024/

// `frame_support` and `frame_system` are declared here
use polkadot_sdk::*;

#[frame_support::pallet(dev_mode)]
pub mod pallet {
    use super::*;
    use frame_support::pallet_prelude::*;
    use frame_system::pallet_prelude::*;

    pub type Balance = u128;

    #[pallet::pallet]
    pub struct Pallet<T>(_);

    #[pallet::config]
    #[pallet::disable_frame_system_supertrait_check]
    pub trait Config: frame_system::Config {}

    #[pallet::storage]
    pub type Balances<T: Config> = StorageMap<_, _, T::AccountId, Balance>;

    impl<T: Config> Pallet<T> {
        pub fn transfer(
            from: T::RuntimeOrigin,
            to: T::AccountId,
            amount: Balance,
        ) -> DispatchResult {
            let sender = ensure_signed(from)?;
            let sender_balance = Balances::<T>::get(&sender).ok_or("NonExistentAccount")?;
            let sender_remainder = sender_balance
                .checked_sub(amount)
                .ok_or("InsufficientBalance")?;

            Balances::<T>::mutate(to, |b| *b = Some(b.unwrap_or(0) + amount));
            Balances::<T>::insert(&sender, sender_remainder);

            Ok(())
        }
    }
}
```

> To learn more about building with the Polkadot SDK, you may start with these [guides](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/guides/index.html).

Reviewer: Let's link to both this, and `docs.polkadot.com`.

## 💡 Pro Tips

(Any suggestions?)

In Substrate, a runtime can be seen as a tuple of various pallets. Since the number of pallets varies and there is no way to anticipate it, we have to generate impl-trait for tuples of different sizes upfront, from 0-tuple to 64-tuple to be specific (64 is chosen to balance usability and compile time).

Seldomly, if the runtime grows larger than 64 pallets, the trait implementations will no longer apply, then the feature `tuples-96` (or even `tuples-128`) is required (at the cost of increased compile time).

> ## 🚀 Versioning

Reviewer: Link to https://github.com/paritytech/release-registry/ would be useful.

![SDK Logo](../docs/images/Polkadot_Logo_Horizontal_Pink_Black.png#gh-light-mode-only)

# `polkadot-sdk`

<!-- markdownlint-disable-next-line MD013 -->
[![StackExchange](https://img.shields.io/badge/StackExchange-Community%20&%20Support-222222?logo=stackexchange)](https://substrate.stackexchange.com/)

</div>

`polkadot-sdk` is an umbrella crate for the [Polkadot SDK](https://github.com/paritytech/polkadot-sdk), in the sense that it is an "umbrella" that encompasses other components. More specifically, it re-exports all other published crates in the SDK, except for the example and fuzzing, plus some of the runtime crates. `polkadot-sdk` aims to be an entry to the Polkadot and Substrate ecosystem and make the SDK more approachable—the entire development environment made available with one dependency. More importantly, it guarantees the right combination of crate versions so that third-party tools for selecting compatible crate versions are no longer necessary.

## 💻 Usage

The re-exported crates are grouped under the following feature sets.

- `node`
- `runtime`
- `runtime-full`
- `experimental`
- `runtime-benchmarks`
- `serde`
- `tuples-96`
- `try-runtime`

> - `with-tracing`

Reviewer: What are these 4 features and when should each be enabled?

Author: Ok, as @ggwpez mentioned earlier that some of these are for power-users (at least not needed when someone first starts out to build with `polkadot-sdk`), maybe it makes sense to explain those in the Pro Tips section?

                .checked_sub(amount)
                .ok_or("InsufficientBalance")?;

            Balances::<T>::mutate(to, |b| *b = Some(b.unwrap_or(0) + amount));
            Balances::<T>::insert(&sender, sender_remainder);

            Ok(())
        }
    }
}
```

To learn more about building with the Polkadot SDK, you may start with these [guides](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/guides/index.html) and our [official docs](https://docs.polkadot.com/).

## 💡 Pro Tips

In Substrate, a runtime can be seen as a tuple of various pallets. Since the number of pallets can  vary and there is no way to anticipate it, we have to generate impl-trait for tuples of different sizes upfront, from 0-tuple to 64-tuple to be specific (64 is chosen to balance between usability and compile time). Seldomly, when the runtime grows to have more than 64 pallets, the trait implementations will cease to apply, then the feature `tuples-96` (or even `tuples-128`) must be enabled (at the cost of increased compile time).

```toml
[dependencies]
polkadot-sdk = { version = "0.12.0", features = ["runtime", "tuples-96"], default-features = false }
```

> ## ⬆️ Updating

Reviewer: Let's note this section is for contributors of polkadot-sdk, not the users thereof :)

Author: Yeah, that makes sense.

<div align="center">

![SDK Logo](../docs/images/Polkadot_Logo_Horizontal_Pink_White.png#gh-dark-mode-only)
![SDK Logo](../docs/images/Polkadot_Logo_Horizontal_Pink_Black.png#gh-light-mode-only)

# `polkadot-sdk`

<!-- markdownlint-disable-next-line MD013 -->
[![StackExchange](https://img.shields.io/badge/StackExchange-Community%20&%20Support-222222?logo=stackexchange)](https://substrate.stackexchange.com/)

</div>

> `polkadot-sdk` is an umbrella crate for the [Polkadot SDK](https://github.com/paritytech/polkadot-sdk), in the sense that it is an "umbrella" that encompasses other components. More specifically, it re-exports all other published crates in the SDK, except for the example, fuzzing, and some of the runtime crates.

Reviewer 2: ```suggestion
`polkadot-sdk` is an umbrella crate for the [Polkadot SDK](https://github.com/paritytech/polkadot-sdk), in the sense that it is an "umbrella" that encompasses other components. More specifically, it re-exports all creates that are needed for builders of runtimes and nodes in a single dependency ✨
```

`polkadot-sdk` aims to be the entry to the Polkadot and Substrate ecosystem and make the SDK more approachable—the entire development environment made available with **one dependency**. More importantly, it guarantees the compatible combination of crate versions. So even if you know exactly which crates to use, you may still benefit from using `polkadot-sdk` for painless dependency updates.

You may have seen another umbrella crate named `polkadot-sdk-frame`. For clarification, while `polkadot-sdk` aims to ease dependency management, `polkadot-sdk-frame` intends to simplify [FRAME](https://docs.polkadot.com/polkadot-protocol/glossary/#frame-framework-for-runtime-aggregation-of-modularized-entities) pallet implementation, as demonstrated in the example below.

## 💻 Usage

The re-exported crates are grouped under the following feature sets.

* Main

>     - `node`

Reviewer 2: ```suggestion
    - `node`: Anything that you need to build a node
```

>     - `runtime`

Reviewer 2: ```suggestion
    - `runtime`: Most things that you need to build a runtime
```

    - `runtime-full`

> * Accessory

Author: Sure! 😄

    use frame::prelude::*;

    pub type Balance = u128;

    #[pallet::pallet]
    pub struct Pallet<T>(_);

    #[pallet::config]
    #[pallet::disable_frame_system_supertrait_check]
    pub trait Config: frame_system::Config {}

    #[pallet::storage]
    pub type Balances<T: Config> = StorageMap<_, _, T::AccountId, Balance>;

    impl<T: Config> Pallet<T> {
        pub fn transfer(
            from: T::RuntimeOrigin,
            to: T::AccountId,
            amount: Balance,
        ) -> DispatchResult {
            let sender = ensure_signed(from)?;
            let sender_balance = Balances::<T>::get(&sender).ok_or("NonExistentAccount")?;
            let sender_remainder = sender_balance
                .checked_sub(amount)
                .ok_or("InsufficientBalance")?;

            Balances::<T>::mutate(to, |b| *b = Some(b.unwrap_or(0) + amount));
            Balances::<T>::insert(&sender, sender_remainder);

            Ok(())
        }
    }
}
```

To learn more about building with the Polkadot SDK, you may start with these [guides](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/guides/index.html) and our [official docs](https://docs.polkadot.com/).

> ## 💡 Pro Tips

Reviewer 2: Maybe also put this into the drop-down spoiler from above. Github supports `<details>` for that.

```toml
[dependencies]
polkadot-sdk = { version = "0.12.0", features = ["runtime", "tuples-96"], default-features = false }
```

In addition to all the features mentioned earlier, each exported crate is feature-gated individually with the name identical to the crate name, to provide fine-grained control over the dependencies. Enabling features like `node` may pull in dependencies that you don't need. As you become more knowledgeable about the SDK, you may consider specifying the exact crate names in the `features` list instead to reduce build time.

## 🚀 Versioning

We do a stable release for the SDK every three months with a version schema reflecting the release cadence, which is tracked in the [release registry](https://github.com/paritytech/release-registry/). At the time of writing, the latest version is `stable2412` (released in 2024 December). To avoid confusion, we will align the versioning of `polkadot-sdk` with the established schema. For instance, the next stable version will be `2503.0.0`.

> ## 📝 Note to Contributors

Reviewer 2: I think these could go into a `MAINTAIN.md`, its not a convention in our repo yet, but this info does not really pertain to the average reader.

Author: How about putting this info somewhere else like `./CONTRIBUTING.md` instead of in this crate? As it is supposed updated by people who're working on other crates.

# `polkadot-sdk`

> 

Reviewer 2: >For sprucing up the README 🎀—logos, badges, maybe a few emojis. Could I borrow some from the repo’s README?

Yes of course. I think having a unified look could even be beneficial. Although the repo readme is probably a bit over the top for our need here 😆

> `polkadot-sdk` is an umbrella crate re-exporting all other published crates, except for the example and fuzzing, plus some of the runtime crates like Rococo. Consider it the entry to the whole Polkadot and Substrate ecosystem.

Reviewer 2: I think the word "umbrella" is not actually that well known in this context. We may need to explain it more first.

You can also omit Rococo, it will be phased out.

`polkadot-sdk` aims to make the SDK more approachable - the entire development environment made available with one dependency. More importantly, it guarantees the right combination of crate versions, thus third-party tools for selecting compatible crate versions are no longer needed.

## Usage

> The re-exported crates are grouped under the following feature sets.

Reviewer 2: I think we dont need to explain all of these features in detail. The big ones like `node` and `runtime` should probably be first and then the other ones can go into a dropdown for details.

Reviewer: So the scenarios are as follows. 

I want to do the following: 

1. I want to build a node with the umbrella crate (`node` feature)
2. I want to build a runtime with the umbrella crate (`runtime` feature)
3. I want to have benchmarking in the runtime
4. I want to use pallet-x, y and z in my runtime

These should be answered in this section.

Author: Yeah, that’s better! 😊

- `std`: for enabling `std` support of the enabled features
- `experimental`: for experimental features
- `node`: for node implementation
- `runtime`: for runtime implementation
- `runtime-benchmarks`: for benchmarking runtimes
- `runtime-full`: for more comprehensive runtime implementation?
- `serde`: for `serde` en/decoding support
- `tuples-96`: required by runtimes with more than 64 pallets
- `try-runtime`: for cli support?
- `with-tracing`: for tracing support

Choose which ones to enable based on the needs. For example, when building a runtime with more than 64 pallets, benchmarking needs and optional tracing, the `Cargo.toml` may contain

```toml
[dependencies]

> polkadot-sdk = { version = "0.12.0", features = ["runtime", "tuples-96"] }

Reviewer 2: The `tuples-96` is only needed for bigger runtimes and is a power-user feature.  
Can explain it in some advanced section.

[features]
runtime-benchmarks = ["polkadot-sdk/runtime-benchmarks"]

> with-tracing = ["polkadot-sdk/with-tracing"]

Reviewer 2: The tracing is also power user. Maybe focus on the minimal viable usage here - just a runtime.

```

```shell
# build enabling runtime-benchmarks and with-tracing
$ cargo build --features "runtime-benchmarks,with-tracing"
```

## Pro Tips

* fine-grained control over dependencies
* any suggestions?

> In addition to the features above, each re-exported crate is feature-gated individually so that one can control exactly which crates to include. In the contrived example below, only the `frame-support` and `frame-system` crates are needed.

Reviewer 2: Some context:

This is done to reduce build time in cases where only a small subset of dependencies is needed. The less dependencies you enable - the faster it builds.

```rust
// `frame_support` and `frame_system` are declared here
use polkadot_sdk::*;

#[frame_support::pallet(dev_mode)]
pub mod pallet {
    use super::*;
    use frame_support::pallet_prelude::*;
    use frame_system::pallet_prelude::*;

    pub type Balance = u128;

    #[pallet::pallet]
    pub struct Pallet<T>(_);

    #[pallet::config]
    #[pallet::disable_frame_system_supertrait_check]
    pub trait Config: frame_system::Config {}

    #[pallet::storage]
    pub type Balances<T: Config> = StorageMap<_, _, T::AccountId, Balance>;

    impl<T: Config> Pallet<T> {
        pub fn transfer(
            from: T::RuntimeOrigin,
            to: T::AccountId,
            amount: Balance,
        ) -> DispatchResult {
            let sender = ensure_signed(from)?;
            let sender_balance = Balances::<T>::get(&sender).ok_or("NonExistentAccount")?;
            let sender_remainder = sender_balance
                .checked_sub(amount)
                .ok_or("InsufficientBalance")?;

            Balances::<T>::mutate(to, |b| *b = Some(b.unwrap_or(0) + amount));
            Balances::<T>::insert(&sender, sender_remainder);

            Ok(())
        }
    }
}
```

## Versioning

> We do a stable release for the SDK quarterly with a version scheme reflecting the release cadence. At the time of writing, the latest version is `stable2412` (released on December 2024). To avoid confusion, we plan to align with the established scheme for the umbrella crate. For example, the next stable version of `polkadot-sdk` will be `2503.0.0`.

Reviewer 2: Am currently asking the release team on what to do here. Either we go with this, or `0.2503` instead.

Author: I feel like 2503.0.0 might have a better chance of being semver-compatible—a lot more than bug fixes could happen in a 3-month window.
